### ISSUE TITLE
New raw data file format

### ISSUE THREAD

**Issue by dneise** [cta-observatory/dragonboard_testbench#22]
Since 02.03.16 we have a set of new raw data files in a new format. 

The format is described in the "Dragon ver.3 Front-end Board Document"  Version **1.5.1.0B** from 23.08.15 pages 16 and 17.

It differs from the previous format. 

This reader should support both raw data file formats, and decide by itself which format a file is. There is a lot of room for error, when letting humans provide a file format to the reader.

**Comment by dneise** [comment]
In 99d92ba770db26fc5d511cbff31bbd3faa5848e8 the normal EventGetter in dragonboard.io can read only the new format. Who is interested might want to check that out and test it.

At the moment I have no time to merge both the old and the new reader together in a clean way. 

However, who urgently needs to read the new file format can checkout this branch and play.

**Comment by Froehliche-Kernschmelze** [comment]
I have tried the new io.py
I get the following error message:

Traceback (most recent call last):
  File "offset_calibration.py", line 195, in <module>
    store_calibrated_data(output_directory, apply_offset_calibration(raw_datafile_directory, calibration_constants_directory))
  File "offset_calibration.py", line 54, in apply_offset_calibration
    for event in EventGenerator(f):
  File "/home/mario/Programs/lst_calibration/New_Data/dragonboard/io.py", line 172, in __next__
    return self.next()
  File "/home/mario/Programs/lst_calibration/New_Data/dragonboard/io.py", line 187, in next
    data = read_data(self.file_descriptor, roi)
  File "/home/mario/Programs/lst_calibration/New_Data/dragonboard/io.py", line 127, in read_data
    array['high'][channel + 1] = data_odd[channel::8]
ValueError: could not broadcast input array from shape (15) into shape (40)

...anyone got an idea?

**Comment by dneise** [comment]
Mario can you tell us, what exactly you did? For Max and me the reader seems to work .. at least enough to use the Viewer ...

**Comment by Froehliche-Kernschmelze** [comment]
I replaced my io.py in mario_offsetcalculation with your io.py in branch iss_22_proposal that you suggested.

**Comment by dneise** [comment]
And after that? Did you maybe also call some program or so? 

Would you mind sharing what exactly you typed into your shell, so that we can repeat it and get the same error and then see where it came from?

**Comment by Froehliche-Kernschmelze** [comment]
I wanted to apply my offset_calibration.py to the data contained in ~/20160229_185742_ExtTrgRand/, whereas my offsets.csv are stored in ~/Offsets. The output directory was determined to be ~/calibrated_output.

This workes pretty well with the "old" data and the "old" io.py. After Dom announced some changes regarding the io.py I replaced my "old" io.py with the new one in iss_22_proposal. You can see the error messages below.

Let me know if you need further information!

mario@Azure:~/Programs/lst_calibration$ python offset_calibration.py /home/mario/Programs/lst_calibration/New_Data/20160229_185742_ExtTrgRand/ /home/mario/Programs/lst_calibration/Offsets/ /home/mario/Programs/lst_calibration/calibrated_output/ 
Traceback (most recent call last):
  File "offset_calibration.py", line 195, in <module>
    store_calibrated_data(output_directory, apply_offset_calibration(raw_datafile_directory, calibration_constants_directory))
  File "offset_calibration.py", line 54, in apply_offset_calibration
    for event in EventGenerator(f):
  File "/home/mario/Programs/lst_calibration/dragonboard/io.py", line 185, in __next__
    return self.next()
  File "/home/mario/Programs/lst_calibration/dragonboard/io.py", line 198, in next
    roi = get_roi(event_size)
  File "/home/mario/Programs/lst_calibration/dragonboard/io.py", line 40, in get_roi
    assert roi.is_integer()
AssertionError

**Comment by dneise** [comment]
I can reproduce that error. 

It happens after a longer time though. Would have been good to say so. Since I was about to Ctrl-C the program after it ran happily for more than 10seconds.

Searching for the reason now.

**Comment by dneise** [comment]
I must express somehow, that I think it is not me who should have found out that the file is broken. Its not so difficult to find out.

**Comment by dneise** [comment]
Okay ... so since I am so angry now... and feel like I am the only one working on this. I pretend this is the case. 

I merged the current offset calculation of mario into this branch `iss_20_proposal` and rewrote it the way I would like it to be.

Then I executed it (took ~16minutes) and had a look at [the results](https://github.com/cta-observatory/dragonboard_testbench/blob/iss_22_proposal/Look_at_new_offset_calib_results.ipynb)

**Comment by Froehliche-Kernschmelze** [comment]
I received the above mentioned error instantly so I could not report it to happen after some time has passed. I'll try to find out why this happens and try to reproduce what you could find out.

**Comment by dneise** [comment]
This happens because the return of `glob.glob("some/path/*.dat")` is not sorted. So this call on different peoples machines returns the file paths in an arbitrary order. Nothing really bad, but inconvenient.

**Comment by dneise** [comment]
Okay, so now we have a reader aka `EventGetter`, which tries to find out by itself, what file format it is looking at. Here is an example:

~~~
In [1]: p1 = "../raw_data_20160229/20160229_191654_ExtTrgRand/Ped_Rand10kHz_No1_IP90.dat"

In [2]: p2 = "../IP192.168.1.1/Pedestal/Ped444442_1.dat"

In [3]: import dragonboard.io as dio

In [4]: dio.guess_version(p1)
Out[4]: 'v5_1_0B'

In [5]: dio.guess_version(p2)
Out[5]: 'v5_1_05'

In [6]: dio.EventGenerator(p1)
Out[6]: 
EventGenerator_v5_1_0B(
path='/home/dneise/LST/raw_data_20160229/20160229_191654_ExtTrgRand/Ped_Rand10kHz_No1_IP90.dat', 
max_events=None)
roi ......: 40
event_size: 1344 byte
#events ..: 100000

In [7]: dio.EventGenerator(p2)
Out[7]: 
EventGenerator_v5_1_05(
path='/home/dneise/LST/IP192.168.1.1/Pedestal/Ped444442_1.dat', 
max_events=None)
roi ......: 1024
event_size: 32816 byte
#events ..: 1000
~~~

I needed to adjust the plotter a little bit, since I changed the interface of the EventGetter a bit.
I needs now a filename and not a filelike object.

So the plotter now loads any `.dat`-file and plots it.... well ... at least I hope so.

**Comment by MaxNoe** [comment]
This was fixed by merging #26
